### PR TITLE
FIX Error when prefix tuning Gemma 4

### PR DESCRIPTION
**Pull request by BenjaminBossan**
There was an issue with applying prefix tuning to Gemma 4 because the model uses different head dimensions for layers that use sliding window attention. As prefix tuning only initializes a single projection matrix that is used for all layers, this would lead to a shape mismatch.

The solution is to "overprovision" the matrix and then slice the prefix down to size of the layer is smaller. This is not quite as parameter efficient as it could be, but the overhead shouldn't be too large.

For robustness, we also skip layers if the matrix is underprovisioned, but we warn about it and raise an error if all layers are skipped.

Alternatively, we could implement one project per layer, each with the right size, like in https://github.com/google-deepmind/gemma/pull/631/. However, this would be a big refactor and also very hard to make backwards compatible with existing checkpoints, so going with the less efficient solution is preferable.

This PR also contains an independent, single line fix to a prefix tuning test that was referencing a non-existing model.

### CODE REVIEW

**Comment by stharrold** [comment]
Tested peft#3205 (head `c020c11c`) against Gemma-4-E2B + SDPA + PrefixTuning on NVIDIA H100 80GB. **The forward pass succeeds.** ✅

**Environment:**
- `peft @ git+https://github.com/huggingface/peft.git@c020c11c397cdf2d66a34dccceab4246517a28c1` (reports `0.19.2.dev0`)
- `transformers==5.7.0`
- `torch==2.5.1+cu124`, CUDA available
- `attn_implementation="sdpa"`, `torch_dtype=torch.bfloat16`, NVIDIA H100 80GB
- `model_id="google/gemma-4-e2b-it"`, `PrefixTuningConfig(num_virtual_tokens=20, prefix_projection=False)`

**Result:**
- Forward returned `loss=15.6689` (finite, no NaN)
- `input shape: torch.Size([1, 400])`
- `logits shape: torch.Size([1, 400, 262144])`
- The skip-when-KV-mismatch logic from this PR kicked in cleanly. From the run output:
  > `UserWarning: Prefix tuning injected into layers [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]; skipped [15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34] due to KV shape mismatch or shared-KV layers.`

**Caveat on shape:** our test input (`"Hello, world." * 100` with `max_length=1632`) tokenized to 400 tokens, not the 1632 the reproducer in #3201 finding 1 used. So the exact `expanded size (1632) must match the existing size (1652)` error didn't surface at this shape — but the forward path that previously crashed for prefix-tuning + SDPA + Gemma-4 combinations now runs end-to-end, and the PR's layer-skip logic is observable. Happy to retest with a longer literal prompt that hits 1632+ tokens if you want exact-shape repro confirmation.

For the other two findings in #3201 (eager-attn `position_ids` overflow at `modeling_gemma4.py:2262`; P-Tuning v2 PLE 1.69 TB OOM), this PR doesn't change behavior — they remain WONTFIX per the upstream consensus, and we're tracking them on synavistra's side via internal triage.

Thanks for the partial-fix PR — happy to test follow-up shapes (`prefix_projection=True`, multi-batch, longer sequence length) if useful for the merge.

**Comment by BenjaminBossan** [comment]
@zucchini-nlp It would be great if you could review the PR.

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -69,6 +69,28 @@
...
 
 
+def _get_layer_kv_target_shape(base_config, layer_idx: int) -> tuple[int, int] | None:
+    """Per-layer (num_kv_heads, head_dim) for prefix-tuning injection, or None for uniform models.
+
+    Models with heterogeneous attention (e.g. Gemma4) expose `global_head_dim` / `num_global_key_value_heads` alongside
+    the sliding-layer `head_dim` / `num_key_value_heads`. The provisioned prefix is sized for the global footprint;
+    this returns the shape each layer actually expects so the caller can slice down or skip layers that don't fit.
+    """
+    layer_types = getattr(base_config, "layer_types", None)
```
Comment (zucchini-nlp) [inline src/peft/peft_model.py:79]: so ig we're supporting specifically gemma4 with hardcoded attr names

File: src/peft/peft_model.py [hunk 2]
```diff
@@ -815,12 +837,54 @@ def get_prompt(
...
+                )
+                injected_layers: list[int] = []
+                skipped_layers: list[int] = []
+                # past_key_values is a tuple of `num_layers` per-layer tensors each shaped
+                # [2, batch, num_heads, num_virtual_tokens, head_dim], where dim 0 stacks K and V.
+                for layer_idx, layer_past_key_values in enumerate(past_key_values):
+                    if num_kv_shared_layers > 0 and layer_idx >= first_kv_shared_layer_idx:
+                        skipped_layers.append(layer_idx)
+                        continue
+                    key_states, value_states = layer_past_key_values
```
Comment (zucchini-nlp) [inline src/peft/peft_model.py:855]: nice, prev gemma3 also used to skip layers, so we shouldn't need a prefix cache for it

File: tests/test_decoder_models.py [hunk 3]
```diff
@@ -1106,3 +1106,85 @@ def test_merge_and_unload_fixes_tie_word_embeddings_config(self):
...
+                dtype=torch.bfloat16,
+            ).to(self.torch_device)
+            config = PrefixTuningConfig(
+                task_type=TaskType.CAUSAL_LM,
+                num_virtual_tokens=20,
+                prefix_projection=False,
+            )
+            model = get_peft_model(model, config)
+            text_config = model.config.get_text_config()
+            text_config.num_key_value_heads = 999
```
Comment (zucchini-nlp) [inline tests/test_decoder_models.py:1186]: curious about this. If configs dim are changed, doesn't it mean that key/value cache will also be a larger tensor?

**Review by zucchini-nlp — APPROVED** [review]
Lgtm if we want to support gemma4. I think there is no uniform standard way for a unique arch, so hardcoding specific config names is fine

File: src/peft/peft_model.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/peft_model.py:79]: Yeah, if there is a more general approach, LMK, otherwise I'm okay with a Gemma-specific solution.

File: tests/test_decoder_models.py (hunk 3 above)
Comment (BenjaminBossan) [reply tests/test_decoder_models.py:1186]: I'm not 100% sure, but I think it works because the model is already initialized and the cache is already created at this point, so changing the config won't affect it. But I haven't checked the full code path.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks a lot for reviewing @zucchini-nlp. I replied to your comments. LMK if there is anything left, otherwise I'll go ahead and merge.

**Comment by zucchini-nlp** [comment]
Yep, all good for me. The thing about bigger head dim seems to be a specific edge case :)
